stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/channel/internal/ChannelLogic.kt

```diff
@@ -394,7 +414,15 @@ internal class ChannelLogic(
         repos.insertMessages(messages)
     }
 
-    internal fun upsertMessage(message: Message) = upsertMessages(listOf(message))
+    internal fun upsertMessage(message: Message) {
+        /* Do not upsert message if the newest messages are not loaded to avoid breaking pagination. If the message
+        * is inside the list and was updated in some kind of way, update it in the list. */
+        if (mutableState._endOfNewerMessages.value ||
+            mutableState._messages.value.containsKey(message.id)
```

Comment:
This change might introduce unexpected behavior and we need to test it carefully.
Also, I'm not sure if we should do this check here:
1. Why shouldn't we upsert one message if it's not present in the state but have different behavior for a list of message?
2. Rather an edge case but what will happen if I already watch a channel (so I can receive events) or send messages but the `_endOfNewerMessages ` value was not updated yet?

Reply:
1. If we click on a quoted message and it load n pages away, the newest messages will not be inside the list. The `_endOfNewerMessages` is here to safeguard that we do not insert any new messages that are received to the list. If we add them to the bottom it would break pagination (n number of pages would be missing). The contains ensures that if the messages is contained in the list, any action that modifies it will be propagated (edit, reactions) and will be visible in the list.
2. This could happen yes, if we are initially loading the list, until the call finishes the boolean will be false. But if we have a boolean that checks whether the initial loading is taking place, we can upsert newer messages.

If this kind of safeguard makes more sense elsewhere we can discuss where to move it.

Reply:
Also we should check where it makes sense to block upserting a list of messages sense to prevent breaking pagination.